src/gc/collector.cpp

```diff
@@ -82,20 +107,18 @@ class TraceStack {
 
 public:
     TraceStack() { get_chunk(); }
-    TraceStack(const std::unordered_set<void*>& rhs) {
-        get_chunk();
-        for (void* p : rhs) {
-            assert(!isMarked(GCAllocation::fromUserData(p)));
-            push(p);
-        }
-    }
+    virtual ~TraceStack() {}
+
+    // Returns false if we should skip visiting the children.
+    // true otherwise.
+    virtual bool visit_action(GCAllocation* al) = 0;
```

Comment:
the `virtual` makes me nervous, considering how stupidly hot the trace stack stuff is.  We can defer it until after the `objects_with_finalizers` pass is merged into the sweep phase, but this might benefit from refactoring the TraceStack and using templates + functors to guarantee everything gets inlined.